File: ruby/lib/resp_bench/config/workload_config.rb
# frozen_string_literal: true

module RespBench
  module Config
    # Configuration for a benchmark workload.
    # Maps to the JSON schema in configs/schemas/workload-config.schema.json
    class WorkloadConfig
      attr_accessor :schema_version, :benchmark_profile, :phases

      def initialize(schema_version:, benchmark_profile:, phases:)
        @schema_version = schema_version
        @benchmark_profile = benchmark_profile
        @phases = phases
      end

      def name
        @benchmark_profile[:name]
      end

      def description
        @benchmark_profile[:description]
      end

      def to_h
        {
          schema_version: @schema_version,
          benchmark_profile: @benchmark_profile,
          phases: @phases.map(&:to_h)
        }
      end
    end
  end
end

File: python/src/resp_bench/config/workload_config.py
"""Configuration for a benchmark workload (a sequence of phases)."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional

from .phase_config import PhaseConfig


@dataclass
class WorkloadConfig:
    schema_version: str
    benchmark_profile: dict = field(default_factory=dict)
    phases: List[PhaseConfig] = field(default_factory=list)

    def name(self) -> Optional[str]:
        return self.benchmark_profile.get("name")

    def description(self) -> Optional[str]:
        return self.benchmark_profile.get("description")
